// deck.gl/dist-es6/controllers/interactive-map.js
var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

import { PureComponent, PropTypes, createElement } from 'react';
import MapControls from './map-controls';

// InteractiveMap is for browser only and should work with isomorphic rendering
// Conditionally require mapbox only in browser to avoid issues under node.js
// Node apps can use StaticMap directly - requires adding additional dependencies
import { isBrowser } from './globals';
var StaticMap = isBrowser ? require('react-map-gl').StaticMap : null;

var propTypes = {
  displayConstraints: PropTypes.object.isRequired
};

var defaultProps = {
  displayConstraints: {
    maxPitch: 60
  }
};

var InteractiveMap = function (_PureComponent) {
  _inherits(InteractiveMap, _PureComponent);

  function InteractiveMap(props) {
    _classCallCheck(this, InteractiveMap);

    return _possibleConstructorReturn(this, (InteractiveMap.__proto__ || Object.getPrototypeOf(InteractiveMap)).call(this, props));
  }

  // TODO - Remove once Viewport alternative is good enough


  _createClass(InteractiveMap, [{
    key: '_getMap',
    value: function _getMap() {
      return this._map._map;
    }

    // Checks a displayConstraints object to see if the map should be displayed

  }, {
    key: 'checkDisplayConstraints',
    value: function checkDisplayConstraints(props) {
      var capitalize = function capitalize(s) {
        return s[0].toUpperCase() + s.slice(1);
      };

      var displayConstraints = props.displayConstraints;

      for (var propName in props) {
        var capitalizedPropName = capitalize(propName);
        var minPropName = 'min' + capitalizedPropName;
        var maxPropName = 'max' + capitalizedPropName;

        if (minPropName in displayConstraints && props[propName] < displayConstraints[minPropName]) {
          return false;
        }
        if (maxPropName in displayConstraints && props[propName] > displayConstraints[maxPropName]) {
          return false;
        }
      }
      return true;
    }
  }, {
    key: 'render',
    value: function render() {
      var _this2 = this;

      // Bail out under node
      if (!StaticMap) {
        return null;
      }

      var mapVisible = this.checkDisplayConstraints(this.props);
      var visibility = mapVisible ? 'visible' : 'hidden';

      return createElement(MapControls, Object.assign({}, this.props, {
        key: 'map-controls',
        style: { position: 'relative' }
      }), [createElement(StaticMap, Object.assign({}, this.props, {
        key: 'map-static',
        style: { position: 'absolute', left: 0, top: 0, visibility: visibility },
        ref: function ref(map) {
          _this2._map = map;
        }
      })), createElement('div', {
        key: 'map-children',
        style: { position: 'absolute', left: 0, top: 0 }
      }, this.props.children)]);
    }
  }]);

  return InteractiveMap;
}(PureComponent);

export default InteractiveMap;


InteractiveMap.displayName = 'InteractiveMap';
InteractiveMap.propTypes = propTypes;
InteractiveMap.defaultProps = defaultProps;
